### PR TITLE
Schnorr2

### PR DESCRIPTION
**Pull request by gandrewstone**

### CODE REVIEW

File: src/script/interpreter.cpp [hunk 1]
```diff
@@ -493,6 +493,26 @@ static bool CheckSignatureEncodingSigHashChoice(const vector<unsigned char> &vch
     {
         return true;
     }
+
+    if (flags & SCRIPT_ENABLE_SCHNORR)
+    {
+        if (vchSig.size() == 64 + (check_sighash == true) ? 1 : 0) // 64 sig length plus 1 sighashtype
```
Comment (gubatron) [inline src/script/interpreter.cpp:499]: would this simplification work?
`if (vchSig.size() == 64 + (check_sighash ) ? 1 : 0)`

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -493,6 +493,26 @@ static bool CheckSignatureEncodingSigHashChoice(const vector<unsigned char> &vch
...
+        {
+            // In a generic-signature context, 64-byte signatures are interpreted
+            // as Schnorr signatures (always correctly encoded) when flag set.
+            if (check_sighash && ((flags & SCRIPT_VERIFY_STRICTENC) != 0))
```
Comment (gubatron) [inline src/script/interpreter.cpp:503]: and this? 
`if (check_sighash && (flags & SCRIPT_VERIFY_STRICTENC))`

_just curious why all the flag checks sometimes compare against something or not, e.g. on line 497 it was_
`if (flags & SCRIPT_ENABLE_SCHNORR)` and not `if ((flags & SCRIPT_ENABLE_SCHNORR) != 0)`

File: src/script/interpreter.cpp [hunk 3]
```diff
@@ -493,6 +493,26 @@ static bool CheckSignatureEncodingSigHashChoice(const vector<unsigned char> &vch
...
+            {
+                if (!IsDefinedHashtypeSignature(vchSig))
+                    return set_error(serror, SCRIPT_ERR_SIG_HASHTYPE);
+
+                // schnorr sigs must use forkid sighash if forkid flag set
+                if ((flags & SCRIPT_ENABLE_SIGHASH_FORKID) && ((vchSig[64] & SIGHASH_FORKID) == 0))
```
Comment (gubatron) [inline src/script/interpreter.cpp:509]: and here maybe:
`if ((flags & SCRIPT_ENABLE_SIGHASH_FORKID) && (!(vchSig[64] & SIGHASH_FORKID)))`

File: src/script/interpreter.cpp [hunk 4]
```diff
@@ -1584,6 +1604,19 @@ bool ScriptMachine::Step()
                         valtype &vchSig = stacktop(-isig);
                         valtype &vchPubKey = stacktop(-ikey);
 
+                        // If schnorr is enabled, then no signature can be 64 + 1 bytes because multisig does
+                        // not support schnorr, and all 64 byte signatures are assumed to be schnorr.
+                        if (flags & SCRIPT_ENABLE_SCHNORR)
+                        {
+                            if (vchSig.size() == 65) // 64 sig length plus 1 sighashtype
```
Comment (gubatron) [inline src/script/interpreter.cpp:1611]: wonder if we should make the sig length a constant, all the 64s and 65s could lead to human error

File: src/script/sigcache.cpp [hunk 5]
```diff
@@ -109,17 +126,71 @@ void InitSignatureCache()
...
+    const F &fun)
+{
+    uint256 entry;
+    signatureCache.ComputeEntry(entry, vchSig, pubkey, sighash, flags);
+    if (signatureCache.Get(entry, !storeOrErase))
+    {
+        return true;
+    }
+
+    if (!fun())
```
Comment (gubatron) [inline src/script/sigcache.cpp:144]: is the code above the call to this passed function so expensive?
or is the code of the passed function so commonly called and expensive sometimes that you rather defer it?, say:

`[&] { return TransactionSignatureChecker::VerifySignature(vchSig, pubkey, sighash); }` is one that I see you pass.


if not, perhaps it's simpler to just pass the bool output of the function as a parameter, even if it's not to be used.
It's a pretty cool optimization, but when I see that sometimes you pass a lambda that returns `false` it's kind of weird.

File: src/secp256k1/src/modules/recovery/main_impl.h [hunk 6]
```diff
@@ -126,6 +126,7 @@ int secp256k1_ecdsa_sign_recoverable(const secp256k1_context* ctx, secp256k1_ecd
     int recid;
     int ret = 0;
     int overflow = 0;
+    const unsigned char secp256k1_ecdsa_recoverable_algo16[17] = "ECDSA+Recovery  ";
```
Comment (gubatron) [inline src/secp256k1/src/modules/recovery/main_impl.h:129]: why size 17 and not 16? (the given string is 16 chars)
```
>>> len("ECDSA+Recovery  ")
16
```

File: src/secp256k1/src/secp256k1.c [hunk 7]
```diff
@@ -360,6 +360,7 @@ int secp256k1_ecdsa_sign(const secp256k1_context* ctx, secp256k1_ecdsa_signature
     secp256k1_scalar sec, non, msg;
     int ret = 0;
     int overflow = 0;
+    const unsigned char secp256k1_ecdsa_der_algo16[17] = "ECDSA+DER       ";
```
Comment (gubatron) [inline src/secp256k1/src/secp256k1.c:363]: same here, given that 17 shows up again, I wonder if you're missing a right padding space in the string
```
>>> len("ECDSA+DER       ")
16
```

File: src/script/sigcache.cpp [hunk 8]
```diff
@@ -109,17 +127,66 @@ void InitSignatureCache()
...
+
+bool CachingTransactionSignatureChecker::IsCached(const std::vector<uint8_t> &vchSig, const CPubKey &pubkey,
+                                                  const uint256 &sighash) const
+{
+    return RunMemoizedCheck(vchSig, pubkey, sighash, nFlags, true, [] { return false; });
+}
+
+
+#if 0 // TODO
 bool CachingTransactionSignatureChecker::VerifySignature(const std::vector<unsigned char> &vchSig,
```
Comment (ptschip) [inline src/script/sigcache.cpp:163]: do we still need to keep this #if 0 section

File: src/script/interpreter.cpp (hunk 2 above)
Comment (gandrewstone) [reply src/script/interpreter.cpp:503]: Its a matter of individual style and experience with other languages I think. Some people prefer to be very explicit (me), others rely on the fact that x & y may evaluate to nonzero and any nonzero value is cast to boolean true, while zero is cast to false.

I don't think that BU has a coding standard on this.

File: src/secp256k1/src/modules/recovery/main_impl.h (hunk 6 above)
Comment (gandrewstone) [reply src/secp256k1/src/modules/recovery/main_impl.h:129]: null terminator

File: src/script/sigcache.cpp (hunk 8 above)
Comment (gandrewstone) [reply src/script/sigcache.cpp:163]: I don't love the technique currently undefed that deploys templates and anonymous functions in a 30 line function, whose purpose is seemingly to save 4 or so repeated lines of code.  I am planning to analyze what's exactly going on here more closely and probably break it back into 2 functions.

But I think we should not let this slow us down.  We should move forward and make these cleanups in the final PR that turns on Schnorr at the fork time.

File: src/script/sigcache.cpp (hunk 5 above)
Comment (gandrewstone) [reply src/script/sigcache.cpp:144]: see my comment on the #ifdefed out below

File: src/script/interpreter.cpp (hunk 1 above)
Comment (gandrewstone) [reply src/script/interpreter.cpp:499]: as below

File: src/script/interpreter.cpp (hunk 3 above)
Comment (gandrewstone) [reply src/script/interpreter.cpp:509]: as above

**Review by ptschip — APPROVED** [review]
Looks like you got the last couple of test bugs :)

**Pull request merged**